### PR TITLE
pr: Implement this command

### LINKED ISSUE #1874: implement pr - convert text files for printing
The goal is to implement "pr" 
The work started here:
https://github.com/uutils/coreutils/pull/1327

I will require some rebasing but it should not be too hard.

### PR DESCRIPTION
**Pull request by MaxSem**
This PR is a resurrection of #1327 by @tilakpatidar.

Closes #1874

### CODE REVIEW

File: src/uu/pr/src/pr.rs [hunk 1]
```diff
@@ -0,0 +1,1343 @@
...
+use std::io::{stdin, stdout, BufRead, BufReader, Lines, Read, Stdin, Stdout, Write};
+use std::iter::{FlatMap, Map};
+use std::num::ParseIntError;
+#[cfg(unix)]
+use std::os::unix::fs::FileTypeExt;
+use std::vec::Vec;
+
+type IOError = std::io::Error;
+
+static NAME: &str = "pr";
```
Comment (sylvestre) [inline src/uu/pr/src/pr.rs:31]: Could you please follow this example to declarations the options?
https://github.com/uutils/coreutils/blob/master/src/uu/ls/src/ls.rs#L78-L80

File: src/uu/pr/src/pr.rs [hunk 2]
```diff
@@ -0,0 +1,1343 @@
...
+
+        NotExists(path: String) {
+            display("pr: cannot open {}, No such file or directory", path)
+        }
+    }
+}
+
+pub fn uumain(args: impl uucore::Args) -> i32 {
+    let args = args.collect_str();
+    let mut opts = getopts::Options::new();
```
Comment (sylvestre) [inline src/uu/pr/src/pr.rs:183]: Not blocker but it would be nice to use clap instead of getopts

File: src/uu/pr/src/pr.rs [hunk 3]
```diff
@@ -0,0 +1,1343 @@
...
+}
+
+/// Returns re-written arguments which are passed to the program.
+/// Removes -column and +page option as getopts cannot parse things like -3 etc
+/// # Arguments
+/// * `args` - Command line arguments
+fn recreate_arguments(args: &Vec<String>) -> Vec<String> {
+    let column_page_option = Regex::new(r"^[-+]\d+.*").unwrap();
+    let num_regex: Regex = Regex::new(r"(.\d+)|(\d+)|^[^-]$").unwrap();
+    //let a_file: Regex = Regex::new(r"^[^-+].*").unwrap();
```
Comment (sylvestre) [inline src/uu/pr/src/pr.rs:451]: please remove useless comments

**Review by sylvestre — COMMENTED** [review]
it will take some time to review it but here is a few comments

File: src/uu/pr/src/pr.rs [hunk 4]
```diff
@@ -0,0 +1,1343 @@
...
+        .collect()
+}
+
+fn print_error(matches: &Matches, err: PrError) {
+    if !matches.opt_present(SUPPRESS_PRINTING_ERROR) {
+        eprintln!("{}", err);
+    }
+}
+
+fn print_usage(opts: &mut Options, matches: &Matches) -> i32 {
```
Comment (sylvestre) [inline src/uu/pr/src/pr.rs:485]: this can be removed when/if moved to clap

File: src/uu/pr/src/pr.rs [hunk 5]
```diff
@@ -0,0 +1,1343 @@
...
+    0
+}
+
+fn parse_usize(matches: &Matches, opt: &str) -> Option<Result<usize, PrError>> {
+    let from_parse_error_to_pr_error = |value_to_parse: (String, String)| {
+        let i = value_to_parse.0;
+        let option = value_to_parse.1;
+        i.parse::<usize>().map_err(|_e| {
+            PrError::EncounteredErrors(format!("invalid {} argument '{}'", option, i))
+        })
```
Comment (sylvestre) [inline src/uu/pr/src/pr.rs:535]: please add a test to cover this

File: src/uu/pr/src/pr.rs [hunk 6]
```diff
@@ -0,0 +1,1343 @@
...
+            String::new()
+        } else {
+            if paths[0] == FILE_STDIN {
+                String::new()
+            } else {
+                paths[0].to_string()
+            }
+        });
+
+    let default_first_number: usize = NumberingMode::default().first_number;
```
Comment (sylvestre) [inline src/uu/pr/src/pr.rs:583]: same (test coverage)

File: tests/common/util.rs [hunk 7]
```diff
@@ -127,6 +127,15 @@ impl CmdResult {
         let contents = read_scenario_fixture(&self.tmpd, file_rel_path);
         self.stdout_is(contents)
     }
```
Comment (sylvestre) [inline tests/common/util.rs:129]: please run rustfmt

File: tests/by-util/test_pr.rs [hunk 8]
```diff
@@ -0,0 +1,513 @@
+use crate::common::util::*;
```
Comment (sylvestre) [inline tests/by-util/test_pr.rs:1]: please run rustfmt

File: src/uu/pr/src/pr.rs [hunk 9]
```diff
@@ -0,0 +1,1343 @@
+#![crate_name = "uu_pr"]
```
Comment (sylvestre) [inline src/uu/pr/src/pr.rs:1]: the file is quite long, maybe it would make sense to split it?

**Review by sylvestre — CHANGES_REQUESTED** [review]
well done!

**Comment by sylvestre** [comment]
@MaxSem ping?

**Comment by MaxSem** [comment]
Sorry, was busy doing something else. Will try to find time in the next few days.

**Comment by sylvestre** [comment]
@MaxSem  ping? :)

**Comment by sylvestre** [comment]
@MaxSem ping? :)

**Comment by sylvestre** [comment]
@MaxSem ping?

**Comment by sylvestre** [comment]
moved to
https://github.com/uutils/coreutils/pull/2300

**Pull request closed without merging**